### ISSUE TITLE
Bug in Bio.KEGG.Compound (BioPython 1.65)

### ISSUE THREAD

**Issue by wegnerce** [biopython/biopython#676]
Hi,
iam currently working a lot with the KEGG API implemented in Biopython.
I have problems running this simple example:

``` python
from Bio.KEGG.REST import kegg_get, kegg_list
from Bio.KEGG import Compound

with open("/home/Desktop/test.txt", "w") as out_handle:
    out_handle.write(kegg_get("C00021").read())
with open("/home/Desktop/test.txt") as in_handle:
    for record in Compound.parse(in_handle):
        print record
```

``` python
 for record in Compound.parse(in_handle):
  File "/home/anaconda2/lib/python2.7/site-packages/Bio/KEGG/Compound/__init__.py", line 179, in parse
    pathway = record.pathway[-1]
IndexError: list index out of range
```

I could imagine that KEGG has maybe changed the format a tiny bit, which in turn
causes the observed trouble.

Best wishes
Calle

**Comment by peterjc** [comment]
@kevinwuhoo, @mdehoon - are you familiar with the KEGG compound files?

Thanks Calle - I can confirm this using the latest code from Git (i.e. Biopython 1.66 with minor changes),

```
>>> from Bio.KEGG.REST import kegg_get
>>> from Bio.KEGG import Compound
>>> for record in Compound.parse(kegg_get("C00021")):
...     print record
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Python/2.7/site-packages/Bio/KEGG/Compound/__init__.py", line 179, in parse
    pathway = record.pathway[-1]
IndexError: list index out of range
```

Looking at the raw data,

```
>>> from Bio.KEGG.REST import kegg_get
>>> print kegg_get("C00021").read()
ENTRY       C00021                      Compound
NAME        S-Adenosyl-L-homocysteine;
            S-Adenosylhomocysteine
FORMULA     C14H20N6O5S
EXACT_MASS  384.1216
MOL_WEIGHT  384.4108
REACTION    R00192 R00193 R00194 R00367 R00380 R00594 R00595 R00596 
            R00597 R00598 R00599 R00600 R00601 R00649 R00650 R00683 
            R00827 R00984 R01153 R01159 R01169 R01188 R01189 R01190 
            R01235 R01239 R01269 R01320 R01349 R01721 R01883 R01942 
            R01972 R01973 R02037 R02056 R02144 R02155 R02157 R02174 
            R02175 R02379 R02384 R02533 R02534 R02572 R02623 R02627 
            R02667 R02671 R02808 R02858 R02859 R02862 R02882 R02910 
            R02917 R02920 R02931 R02961 R02982 R03003 R03111 R03112 
            R03130 R03194 R03229 R03304 R03366 R03410 R03411 R03424 
            R03456 R03554 R03571 R03587 R03623 R03633 R03701 R03721 
            R03725 R03731 R03746 R03788 R03805 R03832 R03835 R03875 
            R03922 R03934 R03938 R03943 R03948 R03950 R03955 R04013 
            R04033 R04062 R04077 R04145 R04169 R04190 R04237 R04301 
            R04332 R04396 R04398 R04421 R04427 R04436 R04481 R04496 
            R04505 R04691 R04692 R04705 R04707 R04711 R04713 R04714 
            R04715 R04716 R04717 R04718 R04719 R04762 R04763 R04764 
            R04772 R04858 R04866 R04867 R04881 R04887 R04905 R04931 
            R04981 R04983 R04988 R04990 R04993 R05035 R05122 R05123 
            R05149 R05162 R05179 R05180 R05181 R05182 R05211 R05212 
            R05213 R05214 R05216 R05219 R05303 R05323 R05520 R05532 
            R05614 R05673 R05674 R05676 R05754 R05755 R05756 R05757 
            R05759 R05760 R05763 R05764 R05776 R05808 R05809 R05810 
            R05811 R05813 R05834 R05881 R05885 R06427 R06432 R06435 
            R06439 R06456 R06476 R06477 R06478 R06479 R06564 R06574 
            R06575 R06576 R06577 R06578 R06627 R06675 R06680 R06687 
            R06693 R06703 R06713 R06715 R06719 R06767 R06771 R06774 
            R06778 R06794 R06807 R06815 R06816 R06824 R06859 R06868 
            R06869 R06998 R07232 R07233 R07234 R07235 R07236 R07237 
            R07238 R07239 R07240 R07241 R07242 R07243 R07244 R07251 
            R07330 R07375 R07481 R07501 R07504 R07521 R07524 R07527 
            R07529 R07533 R07535 R07722 R07724 R07750 R07773 R07774 
            R07775 R07917 R07919 R07920 R07921 R07922 R08151 R08236 
            R08239 R08246 R08410 R08441 R08446 R08449 R08455 R08465 
            R08467 R08468 R08469 R08479 R08537 R08538 R08547 R08769 
            R08771 R08774 R08781 R08787 R08788 R08789 R08791 R08794 
            R08803 R08804 R08932 R08973 R08987 R09062 R09063 R09073 
            R09177 R09189 R09191 R09200 R09369 R09370 R09371 R09523 
            R09543 R09569 R09611 R09612 R09613 R09678 R09679 R09736 
            R09737 R09791 R09792 R09821 R09822 R09823 R09856 R09870 
            R09871 R09872 R09939 R09940 R10060 R10061 R10062 R10166 
            R10167 R10168 R10169 R10195 R10196 R10197 R10198 R10199 
            R10250 R10251 R10252 R10306 R10307 R10403 R10404 R10431 
            R10447 R10448 R10449 R10450 R10451 R10452 R10454 R10458 
            R10472 R10491 R10492 R10495 R10497 R10498 R10523 R10524 
            R10586 R10587 R10588 R10605 R10606 R10645 R10647 R10649 
            R10651 R10652 R10654 R10658 R10680 R10709 R10710 R10716 
            R10806 R10812 R10824 R10956 R10957 R10958 R10959 R10962 
            R10963 R10964 R11020
PATHWAY     map00270  Cysteine and methionine metabolism
            map01100  Metabolic pathways
            map01230  Biosynthesis of amino acids
MODULE      M00035  Methionine degradation
            M00609  Cysteine biosynthesis, methionine => cysteine
ENZYME      1.16.1.8        2.1.1.1         2.1.1.2         2.1.1.4         
            2.1.1.6         2.1.1.7         2.1.1.8         2.1.1.9         
            2.1.1.10        2.1.1.11        2.1.1.12        2.1.1.15        
            2.1.1.16        2.1.1.17        2.1.1.18        2.1.1.20        
            2.1.1.22        2.1.1.25        2.1.1.26        2.1.1.27        
            2.1.1.28        2.1.1.33        2.1.1.34        2.1.1.35        
            2.1.1.37        2.1.1.38        2.1.1.39        2.1.1.40        
            2.1.1.41        2.1.1.42        2.1.1.43        2.1.1.44        
            2.1.1.47        2.1.1.49        2.1.1.50        2.1.1.53        
            2.1.1.55        2.1.1.56        2.1.1.57        2.1.1.59        
            2.1.1.60        2.1.1.61        2.1.1.62        2.1.1.64        
            2.1.1.65        2.1.1.67        2.1.1.68        2.1.1.69        
            2.1.1.70        2.1.1.71        2.1.1.72        2.1.1.75        
            2.1.1.76        2.1.1.77        2.1.1.78        2.1.1.79        
            2.1.1.80        2.1.1.82        2.1.1.83        2.1.1.84        
            2.1.1.85        2.1.1.87        2.1.1.88        2.1.1.89        
            2.1.1.91        2.1.1.94        2.1.1.95        2.1.1.96        
            2.1.1.97        2.1.1.98        2.1.1.99        2.1.1.100       
            2.1.1.101       2.1.1.102       2.1.1.103       2.1.1.104       
            2.1.1.105       2.1.1.106       2.1.1.107       2.1.1.108       
            2.1.1.109       2.1.1.110       2.1.1.111       2.1.1.112       
            2.1.1.113       2.1.1.114       2.1.1.115       2.1.1.116       
            2.1.1.117       2.1.1.118       2.1.1.119       2.1.1.120       
            2.1.1.121       2.1.1.122       2.1.1.123       2.1.1.124       
            2.1.1.125       2.1.1.126       2.1.1.127       2.1.1.128       
            2.1.1.129       2.1.1.130       2.1.1.131       2.1.1.132       
            2.1.1.133       2.1.1.136       2.1.1.137       2.1.1.139       
            2.1.1.140       2.1.1.141       2.1.1.142       2.1.1.143       
            2.1.1.144       2.1.1.145       2.1.1.146       2.1.1.147       
            2.1.1.150       2.1.1.151       2.1.1.152       2.1.1.153       
            2.1.1.154       2.1.1.155       2.1.1.156       2.1.1.157       
            2.1.1.158       2.1.1.159       2.1.1.160       2.1.1.161       
            2.1.1.162       2.1.1.163       2.1.1.164       2.1.1.165       
            2.1.1.169       2.1.1.171       2.1.1.172       2.1.1.173       
            2.1.1.174       2.1.1.175       2.1.1.181       2.1.1.182       
            2.1.1.183       2.1.1.184       2.1.1.187       2.1.1.188       
            2.1.1.195       2.1.1.196       2.1.1.197       2.1.1.201       
            2.1.1.202       2.1.1.203       2.1.1.204       2.1.1.210       
            2.1.1.212       2.1.1.222       2.1.1.228       2.1.1.231       
            2.1.1.232       2.1.1.234       2.1.1.235       2.1.1.236       
            2.1.1.237       2.1.1.238       2.1.1.239       2.1.1.240       
            2.1.1.241       2.1.1.243       2.1.1.254       2.1.1.255       
            2.1.1.261       2.1.1.262       2.1.1.263       2.1.1.267       
            2.1.1.270       2.1.1.271       2.1.1.273       2.1.1.274       
            2.1.1.275       2.1.1.276       2.1.1.277       2.1.1.278       
            2.1.1.279       2.1.1.281       2.1.1.282       2.1.1.283       
            2.1.1.284       2.1.1.285       2.1.1.286       2.1.1.287       
            2.1.1.288       2.1.1.289       2.1.1.290       2.1.1.291       
            2.1.1.292       2.1.1.293       2.1.1.294       2.1.1.295       
            2.1.1.296       2.1.1.297       2.1.1.298       2.1.1.300       
            2.1.1.302       2.1.1.303       2.1.1.304       2.1.1.305       
            2.1.1.306       2.1.1.307       2.1.1.-         2.3.1.161       
            2.3.1.231       2.8.4.3         2.8.4.4         2.8.4.5         
            2.-.-.-         3.2.2.9         3.3.1.1         3.5.4.28        
            6.3.2.40
DBLINKS     CAS: 979-92-0
            PubChem: 3323
            ChEBI: 16680
            KNApSAcK: C00007230
            PDB-CCD: SAH
            3DMET: B01134
            NIKKAJI: J14.397I
ATOM        26
            1   C1y C    30.8464  -22.2148
            2   O2x O    29.7843  -21.4478
            3   C1y C    30.4569  -23.4186
            4   C1y C    28.7458  -22.1913
            5   C1y C    29.1529  -23.4186
            6   O1a O    31.2181  -24.4511
            7   C1b C    27.5303  -21.8136
            8   O1a O    28.3514  -24.5328
            9   N4y N    32.2057  -21.7705
            10  C8y C    33.3358  -22.5885
            11  C8y C    34.4631  -21.7664
            12  N5x N    34.0296  -20.4403
            13  C8x C    32.6344  -20.4428
            14  N5x N    33.4847  -23.9805
            15  C8x C    34.7647  -24.5477
            16  N5x N    35.8919  -23.7256
            17  C8y C    35.7430  -22.3335
            18  N1a N    36.8933  -21.4944
            19  S2a S    26.2976  -22.5400
            20  C1b C    25.0851  -21.8400
            21  C1b C    23.8727  -22.5400
            22  C1c C    22.6603  -21.8400
            23  C6a C    21.4478  -22.5400
            24  O6a O    20.2354  -21.8400
            25  N1a N    22.6603  -20.4401
            26  O6a O    21.4478  -23.9397
BOND        28
            1     1   2 1
            2     1   3 1
            3     2   4 1
            4     3   5 1
            5     3   6 1 #Down
            6     4   7 1 #Up
            7     5   8 1 #Down
            8     4   5 1
            9     1   9 1 #Up
            10    9  10 1
            11   10  11 2
            12   11  12 1
            13   12  13 2
            14    9  13 1
            15   10  14 1
            16   14  15 2
            17   15  16 1
            18   16  17 2
            19   11  17 1
            20   17  18 1
            21    7  19 1
            22   19  20 1
            23   20  21 1
            24   21  22 1
            25   22  23 1
            26   23  24 1
            27   22  25 1 #Up
            28   23  26 2
///
```

The error occurs in parsing the ``PATHWAY`` section,

```
PATHWAY     map00270  Cysteine and methionine metabolism
            map01100  Metabolic pathways
            map01230  Biosynthesis of amino acids
```

This is clearly a little different from the test cases under ``Tests/KEGG/compound.*`` which sadly are not multi-line, but use:

```
PATHWAY     PATH: MAP00860  Porphyrin and chlorophyll metabolism
```

and:

```
PATHWAY     PATH: map00622  Toluene and xylene degradation
```

My guess is the old style ``PATH:`` text is no longer used, whereas previously if I am reading our code correctly it played a role in interpreting multi-line entries?

**Comment by peterjc** [comment]
I think this would be fixed by #1182

**Comment by Gasta88** [comment]
Apologies to bring this issue up again, but even after the merge of #1182 the `test_KEGG.py` fails for the `CompoundTests` section again.

**Comment by peterjc** [comment]
@Gasta88 can you give more details please? What fails and how?

The example failing for me earlier now mostly works - giving the same output on Python 2.7 or 3.6:

```
>>> from Bio.KEGG.REST import kegg_get
>>> from Bio.KEGG import Compound
>>> for record in Compound.parse(kegg_get("C00021")):
...     print(record)
... 
ENTRY       C00021
NAME        S-Adenosyl-L-homocysteine
            S-Adenosylhomocysteine
FORMULA     C14H20N6O5S
PATHWAY     PATH  map00270
            PATH  map01100
            PATH  map01230
ENZYME      1 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              2 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
                      2 (.)           2 (.)           2 (.)           2 (.)    
                  2 (.)           2 (.)           2 (.)           2 (.)        
              3 (.)           3 (.)           3 (.)           6 (.)           
DBLINKS     CAS: 979-92-0
            PubChem: 3323
            ChEBI: 16680
            ChEMBL: CHEMBL142828 CHEMBL418052
            KNApSAcK: C00007230
            PDB-CCD: SAH SAO
            3DMET: B01134
            NIKKAJI: J14.397I
///
```

Note there is something odd/broken with the Enzyme parsing.

**Comment by Gasta88** [comment]
If I run the `test_KEGG.py` script in Python 2 and 3 I get the following:
```
~/Documents/GitHub/biopython/Tests (KEGGUnittest)$ python test_KEGG.py
test_irregular (__main__.CompoundTests) ... ERROR
test_sample (__main__.CompoundTests) ... ERROR
test_4letter (__main__.EnzymeTests) ... ok
test_exceptions (__main__.EnzymeTests) ... ok
test_irregular (__main__.EnzymeTests) ... ok
test_new (__main__.EnzymeTests) ... ok
test_sample (__main__.EnzymeTests) ... ok
test_map00950 (__main__.MapTests) ... ok

======================================================================
ERROR: test_irregular (__main__.CompoundTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_KEGG.py", line 125, in test_irregular
    records = list(Compound.parse(handle))
  File "/home/gasta/.local/lib/python2.7/site-packages/Bio/KEGG/Compound/__init__.py", line 180, in parse
    pathway = record.pathway[-1]
IndexError: list index out of range

======================================================================
ERROR: test_sample (__main__.CompoundTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_KEGG.py", line 109, in test_sample
    records = list(Compound.parse(handle))
  File "/home/gasta/.local/lib/python2.7/site-packages/Bio/KEGG/Compound/__init__.py", line 180, in parse
    pathway = record.pathway[-1]
IndexError: list index out of range

----------------------------------------------------------------------
Ran 8 tests in 0.173s
```

In more details:
```
>>> from Bio.KEGG import Compound
>>> with open("/home/gasta/Documents/GitHub/biopython/Tests/KEGG/compound.sample") as handle:
...     records = list(Compound.parse(handle))
...     print(records[-1].__str__().replace(" ","").split("\n")[:10])
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/gasta/.local/lib/python2.7/site-packages/Bio/KEGG/Compound/__init__.py", line 180, in parse
    pathway = record.pathway[-1]
IndexError: list index out of range
```
for Python 2. Same for Python 3.

The output from your example is the same as mine. 

I've made sure to update my forked repository before testing anything.

**Comment by peterjc** [comment]
The fact that these tests are working (passing) on TravisCI and AppVeyor strongly suggests something is not quite right on your local machine. My guess is something simple like you are not testing with the official master branch, or what is currently installed does not match the git checkout.

In this situation I have often had to clear my local Biopython installation manually, e.g. something like this might work:

```
$ rm -rf ~/.local/lib/python2.7/site-packages/Bio ~/.local/lib/python2.7/site-packages/BioSQL ~/.local/lib/python2.7/site-packages/biopython*
```

**Comment by Gasta88** [comment]
Thanks, it works now.

**Comment by peterjc** [comment]
@kozo2 can you see why this prints something strange for the enzyme lines?

```
>>> from Bio.KEGG.REST import kegg_get
>>> from Bio.KEGG import Compound
>>> for record in Compound.parse(kegg_get("C00021")):
...     print(record)
...
ENTRY       C00021
NAME        S-Adenosyl-L-homocysteine
            S-Adenosylhomocysteine
FORMULA     C14H20N6O5S
PATHWAY     PATH  map00270
            PATH  map01100
            PATH  map01230
ENZYME      1 (.)           2 (.)           2 (.)           2 (.)           2
            (.)           2 (.)           2 (.)           2 (.)           2 (.)
...
```

Thanks!

**Comment by peterjc** [comment]
Thanks @kozo2 for the fix in #1365 for the enzyme output, closing this issue.

@wegnerce these fixes will be included in Biopython 1.71. If you need them urgently you would have to install the fixes from git - please let us know if you need any further assistance, and thank you for reporting this.

**Comment by wegnerce** [comment]
Thanks for the update and the fix.
